pkg/datapath/linux/node.go

```diff
@@ -431,23 +431,100 @@ func (n *linuxNodeHandler) NodeUpdate(oldNode, newNode node.Node) error {
 	return nil
 }
 
-func (n *linuxNodeHandler) enableIPsec(newNode *node.Node) {
+func upsertIPsecLog(err error, spec string, loc, rem *net.IPNet, spi uint8) {
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.Reason: spec,
+		"local-ip":       loc,
+		"remote-ip":      rem,
+		"spi":            spi,
+	})
+	if err != nil {
+		scopedLog.WithError(err).Error("IPsec enable failed")
+	}
+}
+
+func (n *linuxNodeHandler) enableCNIIPsec(v4CIDR []net.IPNet, v6CIDR []net.IPNet) {
 	var spi uint8
 	var err error
-	upsertIPsecLog := func(err error, spec string, loc, rem *net.IPNet, spi uint8) {
-		scopedLog := log.WithFields(logrus.Fields{
-			logfields.Reason: spec,
-			"local-ip":       loc,
-			"remote-ip":      rem,
-			"spi":            spi,
-		})
-		if err != nil {
-			scopedLog.WithError(err).Error("IPsec enable failed")
+
+	n.replaceHostRules()
+
+	for _, cidr := range v4CIDR {
+		ipsecIPv4Wildcard := &net.IPNet{IP: net.ParseIP(wildcardIPv4), Mask: net.IPv4Mask(0, 0, 0, 0)}
+
+		n.replaceNodeIPSecInRoute(&cidr)
+		spi, err = ipsec.UpsertIPsecEndpoint(&cidr, ipsecIPv4Wildcard, ipsec.IPSecDirIn)
+		upsertIPsecLog(err, "CNI In IPv4", &cidr, ipsecIPv4Wildcard, spi)
+
+		n.replaceNodeIPSecOutRoute(&cidr)
+		spi, err = ipsec.UpsertIPsecEndpoint(ipsecIPv4Wildcard, &cidr, ipsec.IPSecDirOut)
+		upsertIPsecLog(err, "CNI Out IPv4", ipsecIPv4Wildcard, &cidr, spi)
+
+		n.replaceNodeExternalIPSecOutRoute(&cidr)
```

Comment:
Is there something IPv4-specific about this call? There's no corresponding call to this in the IPv6 case below.

Reply:
Its there,
```
+       for _, cidr := range v6CIDR {
+               ipsecIPv6Wildcard := &net.IPNet{IP: net.ParseIP(wildcardIPv6), Mask: net.CIDRMask(0, 0)}
+
+               n.replaceNodeIPSecInRoute(&cidr)
+               spi, err = ipsec.UpsertIPsecEndpoint(&cidr, ipsecIPv6Wildcard, ipsec.IPSecDirIn)
+               upsertIPsecLog(err, "CNI In IPv6", &cidr, ipsecIPv6Wildcard, spi)
+
+               n.replaceNodeExternalIPSecOutRoute(&cidr)
+               spi, err := ipsec.UpsertIPsecEndpoint(ipsecIPv6Wildcard, &cidr, ipsec.IPSecDirOut)
+               upsertIPsecLog(err, "CNI Out IPv6", &cidr, ipsecIPv6Wildcard, spi)
+
+       }
+}
```

In a different order though. The missing piece is the replaceNodeIPSecOutRoute()

Reply:
The difference is one is a route to .PrimaryExternal and the other is a route to .Router

Reply:
Ah, okay thanks. Hard to tell the difference between the cases.